Add explicit criteria for merging PRs

camillevilla opened the pull request:
Per discussion in the code4lib Slack with @jrochkind @mejackreed @tpendragon @dkinzer @magibney @cbeer @cdmo @corylown.

Upvotes, downvotes, suggested changes, etc. invited.

dkinzer commented:
This looks fair to me.  I've been following the discussion in the slack channel and I'm conflicted.  I think it's difficult to strike the right balance between the values that awoods mentioned on their slack post:

> Is moving code fast the project/culture priority?
> Is stability and consensus the priority?

I like the idea of stability and consensus but I agree with @jrochkind that the current state is too slow and likely detrimental to the project.

mejackreed commented on CONTRIBUTING.md:
```diff
@@ -18,6 +18,11 @@ Let us know you're interested in working on a feature by filing a ticket in our
+## Pull Requests and Code Review
+- Pull requests must be reviewed and approved by one non-author with write-permissions
+- A reviewer who has approved a PR but has not merged it should provide a comment (e.g. "This looks good, I'm just waiting for Travis to build", "This is a breaking change. I would like at least one more reviewer.")
+- An approved PR may be merged by the reviewer, the author, or another committer
```
My preference is to codify the current working practice as I understand over the past 6 years I have been involved with process. If a reviewer is to approve the pull request, I hope that they could take ownership over that process or ask for more assistance to see it through.

```suggestion
- An approved PR may be merged by the reviewer or another committer
```

mejackreed replied:
Not a blocking comment/suggestion just an opinion ^^

mejackreed replied:
I'm going to remove this suggestion for now in favor moving @camillevilla's PR forward. Hopefully this can be discussed more.

mejackreed commented:
Thanks @camillevilla for pushing this forward!

cdmo commented on CONTRIBUTING.md:
```diff
@@ -18,6 +18,11 @@ Let us know you're interested in working on a feature by filing a ticket in our
 * Submit your change as a [Pull Request](https://help.github.com/en/articles/about-pull-requests).
 
+## Pull Requests and Code Review
+- Pull requests must be reviewed and approved by one non-author with write-permissions
```
```suggestion
- Pull requests must be reviewed and approved by someone who isn't the author
```

camillevilla replied:
I'd prefer to keep this language. See: https://github.com/projectblacklight/blacklight/pull/2116#issuecomment-506525886

cdmo replied:
Works for me, I do appreciate the clarity. 👍

jrochkind commented:
I like it, I think you've done a good job of capturing the baseline consensus community norms, without trying to add too much process above that. 

I *do* agree with @dkinzer that we need to be careful about adding anything that will slow things down more, or add extra required work for someone trying to submit a PR to do. 

As it's written, I don't think this does that. It is basically just "you need one additional set of eyes", which I think is appropriate. I like that it even says it doesn't matter who presses the "merge" button, as long as it has the approval from another set of eyes. "An approved PR may be merged by the reviewer, the author, or another committer"

We can always add more process/stoppers later, if there are problems. These seems the minimal to start with, which seems right -- even if it would speed things up, allowing merge without another set of eyes is not safe. 

We should of course keep in mind that we can all still use our discretion to do *more* than this. If you as an author or reviewer think a change has significant enough consequences that you want more feedback before merging, feel free to try to get it before merging! The rules should not be considered a substitute for discretion. 

 Thanks!

jrochkind commented:
POSSIBLE additions: 

* If code changes are made after approval, another approval (from the same or another qualified person) is required. 

* Discussion from others is always welcome. For changes with significant consequences, more discussion and approval is probably wise, but we know it can be hard to get engagement and if you've tried and can't get it, sometimes we must go on.  Everyone should use their judgement as to trying to get more discussion, and deciding to go on when that has failed.

magibney commented:
>I like that it even says it doesn't matter who presses the "merge" button, as long as it has the approval from another set of eyes. "An approved PR may be merged by the reviewer, the author, or another committer"

I'm inclined to also like this ^^.

I'll point out that this would actually nicely complement a "review window" concept, which as long as set to something reasonably timely (3 days? 1 week?) would provide space for discussion/review. Leaving open the possibility of merging one's own PR (with sufficient approval and an elapsed review period) would leave initiative in the court of the person most directly interested in the PR, which could be appropriate in many cases, and could help reduce the likelihood of PRs languishing.

camillevilla commented:
### Re: @cdmo, Who can approve?
I make approval from a "non-author with write permissions" explicit here because an author might see an approval and get confused about why their PR hasn't been merged. 

Non-author approvals are definitely invited and appreciated (see below).

### Re: @jroch, Code changes
Agreed, it's better to make this policy more explicit. Added in commit (c98e68d5).

### Re: @cdmo @jrochkind Inviting feedback from non-maintainers
Agreed, we want to welcome this. I've added a little language that incorporates this in 647bbe72. I have left out some of the language that @jrochkind wrote re: pragmatism and the overall health of the project because I'm not comfortable owning that proposal and would like to move forward with what we have.

### Re: @magibney, Review windows
I would also prefer to have review windows, but the responses on Slack were either lukewarm or actively against the idea. Maybe the onus of code quality / project health can move to testing release candidates instead. 

### Conclusion / next steps
- I also added a commit reminding people that we hang out on Slack (81b329c1)
- [ ] Personally, I'd be interested in scripting up a bot that handles this specific case: 
>A reviewer who has approved a PR but has not merged it should provide a comment
and comments asking for more context.
- [ ] As a community, we should schedule time to review this both at Blacklight summit and open an asynchronous review period for folks who will not be able to travel
- We often invoke the Samvera community as our main example of another open source software community. Reviewing this documentation again at Blacklight Summit should be a priority. I invite folks who are interested in thinking about this topic to take a look at @ifmeorg's [Code Review Practices](https://github.com/ifmeorg/ifme/wiki/Code-Review-Practices) for inspiration. if-me (another open source project I work on) has a stronger focus on welcoming contributors who are new to open source AND new to coding, so much of the etiquette that developers are socialized into are made more explicit there. No need to discuss in the space of this PR, just an invitation to ponder.

corylown commented:
@camillevilla -- we'll set aside time to review this documentation and related issues at Blacklight Summit.
